pallets/sponsorship/src/migration.rs

```diff
@@ -0,0 +1,342 @@
+	use frame_support::storage_alias;
+	use sp_runtime::traits::Saturating;
+
+	#[derive(Encode, Decode, Debug)]
+	pub struct PotDetails<AccountId, SponsorshipType, Balance: frame_support::traits::tokens::Balance> {
+		pub sponsor: AccountId,
+		pub sponsorship_type: SponsorshipType,
+		pub fee_quota: LimitedBalance<Balance>,
+		pub reserve_quota: LimitedBalance<Balance>,
+	}
+
+	#[derive(Encode, Decode, Debug)]
+	pub struct UserDetails<AccountId, Balance: frame_support::traits::tokens::Balance> {
+		pub proxy: AccountId,
+		pub fee_quota: LimitedBalance<Balance>,
+		pub reserve_quota: LimitedBalance<Balance>,
+	}
+
+	pub type PotDetailsOf<T> =
+		PotDetails<<T as frame_system::Config>::AccountId, <T as Config>::SponsorshipType, BalanceOf<T>>;
+	pub type UserDetailsOf<T> = UserDetails<<T as frame_system::Config>::AccountId, BalanceOf<T>>;
+
+	#[storage_alias]
+	/// Details of a pot.
+	type Pot<T: Config> = StorageMap<Pallet<T>, Blake2_128Concat, <T as Config>::PotId, PotDetailsOf<T>, OptionQuery>;
+
+	#[storage_alias]
+	/// User details of a pot.
+	type User<T: Config> = StorageDoubleMap<
+		Pallet<T>,
+		Blake2_128Concat,
+		<T as Config>::PotId,
+		Blake2_128Concat,
+		<T as frame_system::Config>::AccountId,
+		UserDetailsOf<T>,
+		OptionQuery,
+	>;
+
+	fn migrate_pot_partially<T: Config>(n: usize, starting_key: Vec<u8>) -> (Option<Vec<u8>>, Weight) {
+		let mut iter = Pot::<T>::iter_from(starting_key);
+
+		let pots = iter
+			.by_ref()
+			.take(n)
+			.map(|(pot, details)| {
+				(
+					pot,
+					V1PotDetailsOf::<T> {
+						sponsor: details.sponsor,
+						sponsorship_type: details.sponsorship_type,
+						fee_quota: details.fee_quota,
+						reserve_quota: details.reserve_quota,
+						deposit: Default::default(),
+					},
+				)
+			})
+			.collect::<Vec<_>>();
+
+		let num_of_pots = pots.len();
+
+		pots.into_iter()
+			.for_each(|(pot, details)| V1Pot::<T>::insert(pot, details));
+
+		log::info!(target: "sponsorship", "migrated {} pots", num_of_pots);
+
+		let weight = T::DbWeight::get().reads_writes(num_of_pots as u64, num_of_pots as u64);
+		if num_of_pots == n {
+			(Some(iter.last_raw_key().to_vec()), weight)
+		} else {
+			(None, weight)
+		}
+	}
+
+	fn migrate_user_partially<T: Config>(n: usize, starting_key: Vec<u8>) -> (Option<Vec<u8>>, Weight) {
+		let mut iter = User::<T>::iter_from(starting_key);
+
+		let users = iter
+			.by_ref()
+			.take(n)
+			.map(|(pot, user, details)| {
+				(
+					pot,
+					user,
+					V1UserDetailsOf::<T> {
+						proxy: details.proxy,
+						fee_quota: details.fee_quota,
+						reserve_quota: details.reserve_quota,
+						deposit: Default::default(),
+					},
+				)
+			})
+			.collect::<Vec<_>>();
+
+		let users_len = users.len();
+
+		users.into_iter().for_each(|(pot, user, details)| {
+			UserRegistrationCount::<T>::mutate(&user, |count| {
+				count.saturating_inc();
+			});
+			V1User::<T>::insert(pot, user, details);
+		});
```

Comment:
I wonder if this can be done in one iter

Reply:
Same as my response to other comment